File: 1845-largest-submatrix-with-rearrangements/1845-largest-submatrix-with-rearrangements.cpp
class Solution {
public:
    int largestSubmatrix(vector<vector<int>>& matrix) {
        int m = matrix.size();   // Number of rows
        int n = matrix[0].size(); // Number of columns
        int ans = 0;  // Variable to store the maximum area

        // Step 1: Convert matrix to column heights
        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                // If the current cell is 1 and it's not in the first row
                if (matrix[row][col] != 0 && row > 0) {
                    matrix[row][col] += matrix[row - 1][col];  // Add previous row's value
                }
            }
            
            // Step 2: Sort the row in descending order to maximize area
            vector<int> currRow = matrix[row];
            sort(currRow.begin(), currRow.end(), greater<int>()); // Sort in descending order

            // Step 3: Calculate the largest possible submatrix area
            for (int i = 0; i < n; i++) {
                int area = currRow[i] * (i + 1);  // Height * Width
                ans = max(ans, area);  // Update maximum area
            }
        }
        
        return ans; // Return the maximum found area
    }
};


/*
[[0,0,1],
 [1,1,1],
 [1,0,1]]


{{0, 0, 1}, 
 {1, 1, 2}, 
 {2, 0, 3}}

*/

File: 1845-largest-submatrix-with-rearrangements/1845-largest-submatrix-with-rearrangements.java
class Solution {
    public int largestSubmatrix(int[][] matrix) {
        int row = matrix.length;
        int col = matrix[0].length;

        // Calculate heights for each column
        for (int i = 1; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (matrix[i][j] == 1) {
                    matrix[i][j] += matrix[i - 1][j];
                }
            }
        }

        int res = 0;
        for (int i = 0; i < row; i++) {
            // Sort the heights in ascending order
            Arrays.sort(matrix[i]);

            // Iterate through the sorted heights
            for (int j = 0; j < col; j++) {
                int height = matrix[i][j];
                int width = col - j;
                res = Math.max(res, height * width);
            }
        }

        return res;        
    }
}


/*
class Solution {
    public int largestSubmatrix(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int ans= 0;

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(matrix[i][j] != 0 && i > 0){
                    matrix[i][j] += matrix[i-1][j];
                }
            }

            // Convert row to Integer[] and sort in descending order
            Integer[] currCol = Arrays.stream(matrix[i])  // Convert int[] to Integer[]
                                     .boxed()
                                     .toArray(Integer[]::new);
            Arrays.sort(currCol, Collections.reverseOrder());  // Sort in descending order


            for(int j = 0; j < m; j++){
                int area = currCol[j] * (j+1);
                ans = Math.max(ans, area);
            }
        }

        return ans;
    }
}
*/